[TransposeOptimizer] Support Unsqueeze/Transpose of input consumed by per-axis DQ

adrianlizarraga opened the pull request:
### Description
Follow-up to: https://github.com/microsoft/onnxruntime/pull/21793

- Support looking past a per-axis DQ to do in-place Unsqueeze/Transpose of initializers
- Support looking past a per-axis DQ to cancel a Transpose or Squeeze.

### Test models
For all test models, the transpose optimizer pushes a Transpose through a Mul's input[0]. The Mul's input[1] is optionally unsqueezed and then transposed.

### I. Test in-place unsqueeze and transpose of per-axis quantized weight
Original model has input[1] with shape (3,)
<details><summary>click to expand model image</summary>
<img src="https://github.com/user-attachments/assets/37b6f60c-77d2-4bd3-8ca2-58dc7c88a304" />
</details>

Optimized model has input[1] with shape (1, 3, 1, 1). The initializer was unsqueezed and transposed in-place.
<details><summary>click expand model image</summary>
<img src="https://github.com/user-attachments/assets/adb72757-a164-400c-bfef-2a05f0e35825" />
</details>

### II. Test canceling existing Squeeze before per-axis DQ
Original model has input[1] that is squeezed.
<details><summary>click expand model image</summary>
<img src="https://github.com/user-attachments/assets/f27e6742-b563-42a9-ad06-bb3178b0ceb8" />
</details>

Optimized model unsqueezed and transposed input[1]. The original squeeze was removed due to the unsqueeze, leaving only the Transpose.
<details><summary>click expand model image</summary>
<img src="https://github.com/user-attachments/assets/e56261d4-eba6-4a9f-847b-dcd33548dd07" />
</details>

### III. Test canceling existing Transpose before per-axis DQ
Original model has input[1] that is transposed.
<details><summary>click expand model image</summary>
<img src="https://github.com/user-attachments/assets/f157e04a-572a-479d-8e3b-cf57954df5c0" />
</details>

Optimized model transposed input[1], thus canceling the existing transpose.
<details><summary>click expand model image</summary>
<img src="https://github.com/user-attachments/assets/63d742ce-3762-4ab2-bdb0-1b507886da9d" />
</details>

### IV. Test QDQ fix-up of Transpose/Unsqueeze for per-axis quantization
Original model has input[1] that can be broadcasted.
<details><summary>click expand model image</summary>
<img src="https://github.com/user-attachments/assets/96c0092c-22ec-486d-882e-e2cb59ffe324" />
</details>

The main transpose optimization loop inserts float32 Unsqueeze and Transpose after the DQ. The qdq fix-up pass inserts new per-axis Q/DQ ops after the inserted nodes.
<details><summary>click expand model image</summary>
<img src="https://github.com/user-attachments/assets/b6f89c11-974d-4b35-922f-11effdf06883" />
</details>


### Motivation and Context
Enables the TransposeOptimizer to support more models with per-axis QDQ nodes. Per-axis quantization can improve model accuracy and is used by EPs like QNN.

adrianlizarraga commented on onnxruntime/core/optimizer/transpose_optimization/onnx_transpose_optimization.cc:
```diff
@@ -1363,7 +1544,7 @@ static int EstimateTransposeValueCost(const api::GraphRef& graph, std::string_vi
         if (dq_input_node->OpType() == "Squeeze") {
           auto squeeze_input_node = graph.GetNodeProducingOutput(dq_input_node->Inputs()[0]);
-          if (squeeze_input_node->OpType() == "Transpose") {
+          if (squeeze_input_node != nullptr && squeeze_input_node->OpType() == "Transpose") {
```
Note: fix nullptr dereference for Squeeze that consumes a graph input (`GetNodeProducingOutput()` returns null). Caught by new unit test TransposeOptimizerTests.CancelSqueezeBeforePerAxisDQ.

skottmckay commented on onnxruntime/core/optimizer/transpose_optimization/onnx_transpose_optimization.cc:
```diff
@@ -245,32 +245,232 @@ static std::unique_ptr<api::NodeRef> GetDQIfProducingValue(const api::GraphRef&
+    return scale_shape;
+  }
+
+  // ValueInfo is not available. Try to get shape from Tensor if scale is a constant.
+  const std::unique_ptr<api::TensorRef> dq_scale_constant = graph.GetConstant(scale_input);
+  if (dq_scale_constant) {
+    return dq_scale_constant->Shape();
+  }
```
Is there a scenario where a constant does not have a NodeArg with shape info? I wouldn't have expected so, so I'm not sure if this code is useful.

adrianlizarraga replied:
You're right. There is no scenario in which a constant would not have a valid NodeArg. The `onnxruntime::Graph` class explicitly adds NodeArgs for all constants. I removed this unnecessary logic.

skottmckay commented on onnxruntime/core/optimizer/transpose_optimization/onnx_transpose_optimization.cc:
```diff
@@ -245,32 +245,232 @@ static std::unique_ptr<api::NodeRef> GetDQIfProducingValue(const api::GraphRef&
+    const auto input0_info = graph.GetValueInfo(inputs[0]);
+    auto input0_rank = input0_info->ShapeRank();
+    if (!input0_rank.has_value() || !NormalizeAndValidateAxis(axis, *input0_rank)) {
+      return std::nullopt;  // Unable to normalize the DQ's axis.
```
Side note: We really need to add some sort of log facility now that these subtle things are influencing the behavior. It was simpler in the beginning as there was no QDQ handling so it was trivial to see why a Transpose got added/blocked.

adrianlizarraga replied:
Yeah, logging would probably help debugging/observability. I left a todo comment.

skottmckay commented on onnxruntime/core/optimizer/transpose_optimization/onnx_transpose_optimization.cc:
```diff
@@ -245,32 +245,232 @@ static std::unique_ptr<api::NodeRef> GetDQIfProducingValue(const api::GraphRef&
+      return std::nullopt;  // Unable to normalize the DQ's axis.
+    }
+
+    int64_t block_size = q_or_dq_node.GetAttributeIntDefault("block_size", 0);
+    if (block_size != 0) {
+      // A block size indicates blocked quantization.
+      quant_info.mode = QuantizationMode::kBlocked;
+      quant_info.norm_axis = axis;
+    } else {
+      // A block size of 0 indicates per-axis quantization.
+      quant_info.mode = QuantizationMode::kPerAxis;
+      quant_info.norm_axis = axis;
+    }
```
```suggestion
    quant_info.norm_axis = axis;

    int64_t block_size = q_or_dq_node.GetAttributeIntDefault("block_size", 0);
    if (block_size != 0) {
      // A block size indicates blocked quantization.
      quant_info.mode = QuantizationMode::kBlocked;
    } else {
      // A block size of 0 indicates per-axis quantization.
      quant_info.mode = QuantizationMode::kPerAxis;
    }
```

adrianlizarraga replied:
Updated

skottmckay commented on onnxruntime/core/optimizer/transpose_optimization/onnx_transpose_optimization.cc:
```diff
@@ -245,32 +245,232 @@ static std::unique_ptr<api::NodeRef> GetDQIfProducingValue(const api::GraphRef&
+  /// <param name="graph">graph</param>
+  /// <param name="new_input">name of new unsqueezed input[0]</param>
+  /// <param name="perm_inv">positive unsqueeze axes used to update the DQ's axis</param>
+  void SetUnsqueezedInput(api::GraphRef& graph, std::string_view new_input,
```
nit: the impl for this and SetTransposedInput look to be the same so maybe we could split out into a helper func? happy to have both these functions calling the helper for the clarify it adds. will reduce binary size by a tiny bit as well.

adrianlizarraga replied:
Sounds good. I split out common logic into a private helper.

skottmckay commented on onnxruntime/core/optimizer/transpose_optimization/onnx_transpose_optimization.cc:
```diff
@@ -245,32 +245,232 @@ static std::unique_ptr<api::NodeRef> GetDQIfProducingValue(const api::GraphRef&
-      // the DQ node during TransposeInputImpl and UnsqueezeInput.
-      auto dq_scale = graph.GetConstant(dq_node->Inputs()[1]);
-      if (!dq_scale || dq_scale->NumElements() != 1) {
+      // Scale input to DQ must be a constant initializer.
+      std::string_view dq_scale_input = dq_inputs[1];
+      auto dq_scale_constant = graph.GetConstant(dq_scale_input);
+      if (!dq_scale_constant) {
+        break;
+      }
+
+      // Zero-point input (if present) must be a constant initializer.
+      if (dq_inputs.size() > 2) {
+        std::string_view dq_zp_input = dq_inputs[2];
+        auto dq_zp_constant = graph.GetConstant(dq_zp_input);
+        if (!dq_zp_constant) {
+          break;
+        }
+      }
```
Do these have to be constant initializers or do we just need to know the rank?

adrianlizarraga replied:
These had to be constant initializers because we tried to permute all DQ inputs for quantized roi/sizes/scales/pads. However, now that we're not handling DQs for this special case, we only need to be able to know the rank of the scales input. I removed this code because the call to `GetQuantizationInfo` checks that we can get the scale input's rank.

skottmckay commented on onnxruntime/core/optimizer/transpose_optimization/onnx_transpose_optimization.cc:
```diff
@@ -245,32 +245,232 @@ static std::unique_ptr<api::NodeRef> GetDQIfProducingValue(const api::GraphRef&
+    return quant_info_.mode;
+  }
+
+  inline void DisconectInput0() {
```
```suggestion
  inline void DisconnectInput0() {
```

adrianlizarraga replied:
Updated

skottmckay commented on onnxruntime/core/optimizer/transpose_optimization/onnx_transpose_optimization.cc:
```diff
@@ -1038,16 +1224,43 @@ static void TransposeInputImpl(api::GraphRef& graph, api::NodeRef& node, size_t
     if (constant->Shape().size() == 1 && constant->Shape()[0] == gsl::narrow_cast<int64_t>(perm.size())) {
-      auto& node_to_update = dq_node ? *dq_node : node;
-      Permute1DConstant(graph, node_to_update, *constant, i, constant_to_modify, perm);
+      // TODO(adrianlizarraga): Remove handling of a DQ node from this special case.
+      // A quantized (roi/scales/sizes) input for Resize or a quantized pads input for Pad would be unlikely.
+      // HandleResize/HandlePad permute these kinds of inputs directly and do not try to call TransposeInput on them.
+      // So, we should not have a DQ to look past at this point, but we can handle it until verified.
```
Should be able to check code coverage from existing unit tests. If nothing hits the code for a dq_node being present we can probably assume it's not required. 

If using VS \onnxruntime\VSCodeCoverage.runsettings can be used. Run the unit tests and it should give nice highlighting of the lines that are/aren't hit.

adrianlizarraga replied:
Thanks! I checked code coverage from existing unit tests, and we never enter this special case with a DQ to look past. So, I replaced this logic with an `assert(!dq_to_look_past)`. I left a comment explaining why the callers to `TransposeInput` will not create this scenario.

adrianlizarraga commented:
Hi @skottmckay and @edgchen1, 
Thank you for the helpful review comments. I believe I addressed all comments. Can you please take another look at your convenience? Thanks.

skottmckay reviewed (approved):
Thanks!
